pnc-core/src/main/java/org/jboss/pnc/core/builder/BuildCoordinator.java

```diff
@@ -153,15 +154,17 @@ void startBuilding(BuildTask buildTask) throws CoreException {
         RepositoryManager repositoryManager = repositoryManagerFactory.getRepositoryManager(RepositoryType.MAVEN);
         BuildDriver buildDriver = buildDriverFactory.getBuildDriver(buildTask.getBuildConfiguration().getEnvironment().getBuildType());
         EnvironmentDriver envDriver = environmentDriverFactory.getDriver(buildTask.getBuildConfiguration().getEnvironment());
-
+        
+        RunningEnvironmentWrapper backupRunningEnvironment = new RunningEnvironmentWrapper();
+        
         configureRepository(buildTask, repositoryManager)
-                .thenCompose(repositoryConfiguration -> setUpEnvironment(buildTask, envDriver, repositoryConfiguration))
+                .thenCompose(repositoryConfiguration -> setUpEnvironment(buildTask, envDriver, repositoryConfiguration, backupRunningEnvironment))
                 .thenCompose(runningEnvironment -> buildSetUp(buildTask, buildDriver, runningEnvironment))
                 .thenCompose(runningBuild -> waitBuildToComplete(buildTask, runningBuild))
                 .thenCompose(completedBuild -> retrieveBuildDriverResults(buildTask, completedBuild))
                 .thenCompose(buildDriverResult -> retrieveRepositoryManagerResults(buildTask, buildDriverResult))
                 .thenCompose(buildResults -> destroyEnvironment(buildTask, buildResults))
-                .handle((buildResults, e) -> storeResults(buildTask, buildResults, e));
+                .handle((buildResults, e) -> storeResults(buildTask, buildResults, backupRunningEnvironment, e));
```

Comment:
I think it cannot be done by passing it through the chain, because if some part fails than the method storeResults is called with buildResults  == null. So it has to be somehow backuped out of the chain.